tuning/data/data_processors.py

```diff
@@ -81,33 +81,102 @@ def load_dataset(
+                raise ValueError(f"Data loading failed: invalid {context}.") from e
+            except datasets.exceptions.DatasetGenerationError as e:
+                context = (
+                    f"builder {builder} and data_dir {data_dir}"
+                    if builder and data_dir
+                    else f"builder {builder}"
+                    if builder
+                    else f"path {path}"
+                )
+                raise ValueError(
+                    f"Failed to generate the dataset from the provided {context}."
+                ) from e
+
         if datafile:
-            files = [datafile]
             loader = get_loader_for_filepath(file_path=datafile)
-        elif datasetconfig:
-            files = datasetconfig.data_paths
-            name = datasetconfig.name
-            # simple check to make sure all files are of same type.
-            extns = [get_extension(f) for f in files]
-            assert extns.count(extns[0]) == len(
-                extns
-            ), f"All files in the dataset {name} should have the same extension"
-            loader = get_loader_for_filepath(file_path=files[0])
-
-        if loader in (None, ""):
-            raise ValueError(f"data path is invalid [{', '.join(files)}]")
+            if loader in (None, ""):
+                raise ValueError(f"data path is invalid [{datafile}]")
+            return _load_dataset(builder=loader, data_files=[datafile])
+
+        data_paths = datasetconfig.data_paths
+        builder = datasetconfig.builder
+        all_datasets = []
+
+        for data_path in data_paths:
+            # CASE 1: User passes directory
+            if os.path.isdir(data_path):  # Checks if path exists and isdirectory
+                # Directory case
+                if builder:
+                    # Load using a builder with a data_dir
+                    dataset = _load_dataset(builder=builder, data_dir=data_path)
+                else:
+                    # Load directly from the directory
+                    dataset = _load_dataset(path=data_path)
+            else:
+                # Non-directory (file, pattern, HF dataset name)
+                # If no builder provided, attempt to infer one
+                effective_builder = (
+                    builder if builder else get_loader_for_filepath(data_path)
+                )
+
+                if effective_builder:
+                    # CASE 2: Files passed with builder. Load using the builder and specific files
+                    dataset = _load_dataset(
+                        builder=effective_builder, data_files=[data_path]
+                    )
+                else:
+                    # CASE 3: User passes files/folder/pattern/HF_dataset which has no builder
+                    dataset = _load_dataset(path=data_path)
+
+            all_datasets.append(dataset)
+
+        # Validate all datasets to have same columns
+        validate_datasets(all_datasets)
```

Comment:
The concatenate_datasets function itself will through an error, isn't it? Isn't it better to do it in the catch function of that. Otherwise it will involve some unnecessary computation, isn't it?

Reply:
Based on [this discussion here](https://ibm-research.slack.com/archives/C07U5DX68EP/p1734624445786499), function `validate_mergeable_datasets` now just logs warnings and `load_dataset` allow concatenation of any datasets.